man/man8/ralsh.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "PUPPET\-RESOURCE" "8" "June 2012" "Puppet Labs, LLC" "Puppet manual"
.
.SH "NAME"
\fBpuppet\-resource\fR \- The resource abstraction layer shell
.
.SH "SYNOPSIS"
Uses the Puppet RAL to directly interact with the system\.
.
.SH "USAGE"
puppet resource [\-h|\-\-help] [\-d|\-\-debug] [\-v|\-\-verbose] [\-e|\-\-edit] [\-H|\-\-host \fIhost\fR] [\-p|\-\-param \fIparameter\fR] [\-t|\-\-types] \fItype\fR [\fIname\fR] [\fIattribute\fR=\fIvalue\fR \.\.\.]
.
.SH "DESCRIPTION"
This command provides simple facilities for converting current system state into Puppet code, along with some ability to modify the current state using Puppet\'s RAL\.
.
.P
By default, you must at least provide a type to list, in which case puppet resource will tell you everything it knows about all resources of that type\. You can optionally specify an instance name, and puppet resource will only describe that single instance\.
.
.P
If given a type, a name, and a series of \fIattribute\fR=\fIvalue\fR pairs, puppet resource will modify the state of the specified resource\. Alternately, if given a type, a name, and the \'\-\-edit\' flag, puppet resource will write its output to a file, open that file in an editor, and then apply the saved file as a Puppet transaction\.
.
.SH "OPTIONS"
Note that any configuration parameter that\'s valid in the configuration file is also a valid long argument\. For example, \'ssldir\' is a valid configuration parameter, so you can specify \'\-\-ssldir \fIdirectory\fR\' as an argument\.
.
.P
See the configuration file documentation at http://docs\.puppetlabs\.com/references/stable/configuration\.html for the full list of acceptable parameters\. A commented list of all configuration options can also be generated by running puppet with \'\-\-genconfig\'\.
.
.TP
\-\-debug
Enable full debugging\.
.
.TP
\-\-edit
Write the results of the query to a file, open the file in an editor, and read the file back in as an executable Puppet manifest\.
.
.TP
\-\-host
When specified, connect to the resource server on the named host and retrieve the list of resouces of the type specified\.
.
.TP
\-\-help
Print this help message\.
.
.TP
\-\-param
Add more parameters to be outputted from queries\.
.
.TP
\-\-types
List all available types\.
.
.TP
\-\-verbose
Print extra information\.
.
.SH "EXAMPLE"
This example uses \fBpuppet resource\fR to return a Puppet configuration for the user \fBluke\fR:
.
.IP "" 4
.
.nf

$ puppet resource user luke
user { \'luke\':
 home => \'/home/luke\',
 uid => \'100\',
 ensure => \'present\',
 comment => \'Luke Kanies,,,\',
 gid => \'1000\',
 shell => \'/bin/bash\',
 groups => [\'sysadmin\',\'audio\',\'video\',\'puppet\']
}
.
.fi
.
.IP "" 0
.
.SH "AUTHOR"
Luke Kanies
.
.SH "COPYRIGHT"
Copyright (c) 2011 Puppet Labs, LLC Licensed under the Apache 2\.0 License


CHANGELOG
3.0.0rc3
===
e619304 (#14761) Add boot, reboot to excludes list for redhat provider
7c1d8c2 (maint) Disable upstart spec test on windows
37df5c4 (#14615) Final stub fixes for the tests.
f11dfd6 (#14615) Fix provider spec tests to work on non gentoo systems
060810e maint: Make it easier to test the service provider
c0cb0fa (Maint) Close filehandles after read in upstart
55505db (Maint) Fix upstart provider to work on Ruby 1.9.3
b0f1d58 Update redhat.spec dependency to include hiera-puppet
6e2f23d Update puppet gem dependencies
d19f93f Update redhat init scripts to use `puppet (agent|queue|master)`
a45942b (#14297) Don't remove tty1 or rsyslog during upstart acceptance
466ab49 (#14297) Define agent in upstart acceptance test
f5d1f74 (#14297) Update upstart provider acceptance tests
3c5ef40 (#14297) Clean up upstart provider and tests
36c41e1 (#14343) Test cases for more forms of variable lookup
6cca061 (#14297) Add acceptance test for upstart provider
b9ed026 (#14297) Update spec tests for upstart provider
91827e8 (#14297) Handle upstart services better
3541db6 (#14343) Lookup in correct inherited scope
2d83e22 (Maint) Added test for fully qualified variables
0788ea2 (#14615) Exclude helperscripts in gentoo service provider
2864d86 (maint) Change Facter dep from exit to raise
2ec9b74 (maint) Make Puppet have hard dependency on Facter 2.0
6c37538 (#3909) Move default certname logic to Puppet::Settings
6be09f0 (#3909) Strip trailing dot from default FQDN

3.0.0rc2
===
0cea47e (#14609) Fix master initialization when running under Passenger
b6c5e9d (#14514) Check for the presence of hiera config
ca5271f (maint) Add data binding Hiera tests

3.0.0rc1
===
a731e15 Update CHANGELOG lib/puppet.rb puppet.spec for 2.7.15rc1
53e0057 Use internal domain name for module tool tests
f4ba59e Use absolute_path in module install spec
ead9d25 (Maint) Color console output when the win32console gem is installed
a76d873 (#14424) Expand path of the target directory
95b75e1 Fix unit test failure in fileserver due to modules
da5361d Maint: De-genericize Puppet Forge references in module face
3b144bf Revert "maint: use relative paths to fixtures."
dae5539 (#14440) Hack to allow help faces to be used as API
e862cad Fix broken man page rake task
b26699a (#10146) `-` is not legal in variable names.
7ce9e17 maint: use relative paths to fixtures.
a7a532c Significantly improve compilation performance when using modules
12c5bda Fix a stub of ansicolor that I missed
ea740f6 Add Puppet::Util::Platform to abstract platform checks
91a4a7d Don't use features to check for color in logging
80fe59d Remove a bad test that can lead to random hangs...
0efb4c2 (#14296) Improve template function error message.
d5e0cdd (#14288) gem provider should tolerate bad lines.
574e4c8 Update CHANGELOG, spec files for 2.7.14
d2b3ba0 Revert "Fix incorrect argument handling for expire in NodeExpirer"
7c7d4b6 Fix #14123 for Windows
1c9bec0 Fix incorrect argument handling for expire in NodeExpirer
7e92a7c Fix filebucket specs on Windows
a713f07 (#8778) Make '' == undef commutative in the DSL
24d4bb5 Revert "(#11004) Fix regex for pkg provider line matching"
34da6f7 (#13651) Use magenta, not purple in Facter::Util::Colors
e37b536 (#11004) Fix regex for pkg provider line matching
9b47110 (#13956) Fix failing spec tests
7e92dcb Just change Puppet settings in tests, rather than stubbing.
ca313dc (maint) Fix destinations_spec failure on windows
cde3bee Just change Puppet settings in tests, rather than stubbing.
6aa8441 (#11004) Support Solaris 11 output in pkg package provider
61e7d3c maint: Fix pkg_spec test
603b368 (#14173) Enforce that filebucket paths must be absolute
b7828ed (#3452) Autorequire user for cron
a47c42d Update CHANGELOG for 2.7.14rc3
402a425 (#14127) Add integration tests for ssh_authorized_key
85ee441 maint: refactor integration specs for ssh_authorized_key
42b20be (#13682) Fix acceptance test failures
dce3c32 (#13682) Rename Puppet::Module::Tool to Puppet::ModuleTool
5117e12 (maint) Ensure every file has a trailing newlines
b4d1c65 (#14127) ssh_authorized_keys grammer fails on blank lines.
25ad6d7 Updating CHANGELOG for Puppet 2.7.14rc2
74e462f (#14060) Fix quoting of commands to interpolate inside the shell.
796a072 Modernize the style of the shell exec provider tests.
9e526c9 (#14060) Fix quoting of commands to interpolate inside the shell.
36d959a (#14123) Puppet shouldn't explode if PATH contains ~nonexistent_user
a009c54 Modernize the style of the shell exec provider tests.
934ec89 Improve deprecation warning for dynamic lookup
c255414 (#14036) Add upstart tests
dc3bb9f (#14036) Handle upstart better
a281c48 (#13645) Add test cases to tagmail_spec
3a4ac60 Updating CHANGELOG, lib/puppet.rb for 2.7.14rc1
f6ac1bb (#13645) Do not open a smtp connection if nothing to report
f40cc71 (#12392): Colorize console output on Windows
0d9e852 Removed duplication of compiling a catalog
d67711d Cleaner test for create_resources doing nothing
2d3cc63 Fixing problem caused by activerecord
f1331ba Put back require 'puppet' in Puppet::Util
79bf4fa (#13299) Fix error message when agent is disabled
5178adc Update puppet.spec for rpm packaging
fb96747 Fixed old log test to match new autoflush behavior
878127f Make sure the log file writable
72d3b56 Default autoflushing of log files to true
9711d50 Created test that shows enc warning problem
d54100d (Maint) Fix order-dependent test failure
f531400 (Maint) Don't assume eventlog gem is installed on Windows
2957ffe Fix tempfile cleanup in specs
e2d6915 Fix communication_error acceptance test
27a5eb7 options[:ca_location] is a string not a symbol. Safer to pull Puppet::SSL::Host.ca_location since it is already set correctly.
c01cac9 (#13204) Don't ignore missing PATH.augnew files
7d27cd2 Use :as_platform in the smf service spec
876785a Use :as_platform in the redhat service spec

2.7.16rc1
===
3a06246 (#8858) Explicitly set SSL peer verification mode.
cd4bee8 Test SSL setup code, not our stubbing of it.

lib/puppet/application/ralsh.rb
require 'puppet'
require 'puppet/application'
require 'facter'

Puppet::Application.new(:ralsh) do

    should_not_parse_config

    attr_accessor :host, :extra_params

    preinit do
        @extra_params = []
        @host = nil
        Facter.loadfacts
    end

    option("--debug","-d")
    option("--verbose","-v")
    option("--edit","-e")

    option("--host HOST","-H") do |arg|
        @host = arg
    end

    option("--types", "-t") do |arg|
        types = []
        Puppet::Type.loadall
        Puppet::Type.eachtype do |t|
            next if t.name == :component
            types << t.name.to_s
        end
        puts types.sort
        exit
    end

    option("--param PARAM", "-p") do |arg|
        @extra_params << arg.to_sym
    end

    command(:main) do
        type = ARGV.shift or raise "You must specify the type to display"
        typeobj = Puppet::Type.type(type) or raise "Could not find type #{type}"
        name = ARGV.shift
        params = {}
        ARGV.each do |setting|
            if setting =~ /^(\w+)=(.+)$/
                params[$1] = $2
            else
                raise "Invalid parameter setting %s" % setting
            end
        end

        if options[:edit] and @host
            raise "You cannot edit a remote host"
        end

        properties = typeobj.properties.collect { |s| s.name }

        format = proc {|trans|
            trans.dup.collect do |param, value|
                if value.nil? or value.to_s.empty?
                    trans.delete(param)
                elsif value.to_s == "absent" and param.to_s != "ensure"
                    trans.delete(param)
                end

                unless properties.include?(param) or @extra_params.include?(param)
                    trans.delete(param)
                end
            end
            trans.to_manifest
        }

        text = if @host
            client = Puppet::Network::Client.resource.new(:Server => @host, :Port => Puppet[:puppetport])
            unless client.read_cert
                raise "client.read_cert failed"
            end
            begin
                # They asked for a single resource.
                if name
                    transbucket = [client.describe(type, name)]
                else
                    # Else, list the whole thing out.
                    transbucket = client.instances(type)
                end
            rescue Puppet::Network::XMLRPCClientError => exc
                raise "client.list(#{type}) failed: #{exc.message}"
            end
            transbucket.sort { |a,b| a.name <=> b.name }.collect(&format)
        else
            if name
                obj = typeobj.instances.find { |o| o.name == name } || typeobj.new(:name => name, :check => properties)
                vals = obj.retrieve

                unless params.empty?
                    params.each do |param, value|
                        obj[param] = value
                    end
                    catalog = Puppet::Resource::Catalog.new
                    catalog.add_resource obj
                    begin
                        catalog.apply
                    rescue => detail
                        if Puppet[:trace]
                            puts detail.backtrace
                        end
                    end

                end
                [format.call(obj.to_trans(true))]
            else
                typeobj.instances.collect do |obj|
                    next if ARGV.length > 0 and ! ARGV.include? obj.name
                    trans = obj.to_trans(true)
                    format.call(trans)
                end
            end
        end.compact.join("\n")

        if options[:edit]
            require 'tempfile'
            # Prefer the current directory, which is more likely to be secure
            # and, in the case of interactive use, accessible to the user.
            tmpfile = Tempfile.new('x2puppet', Dir.pwd)
            begin
                # sync write, so nothing buffers before we invoke the editor.
                tmpfile.sync = true
                tmpfile.puts text

                # edit the content
                system(ENV["EDITOR"] || 'vi', tmpfile.path)

                # ...and, now, pass that file to puppet to apply.  Because
                # many editors rename or replace the original file we need to
                # feed the pathname, not the file content itself, to puppet.
                system('puppet -v ' + tmpfile.path)
            ensure
                # The temporary file will be safely removed.
                tmpfile.close(true)
            end
        else
            puts text
        end
    end

    setup do
        Puppet::Util::Log.newdestination(:console)

        # Now parse the config
        Puppet.parse_config

        if options[:debug]
            Puppet::Util::Log.level = :debug
        elsif options[:verbose]
            Puppet::Util::Log.level = :info
        end
    end
end
